### PR TITLE
FACT-539 - Prevent duplicated opening hours types being adding to admin portal

### PR DESCRIPTION
**Pull request by KianKwa**
### JIRA link ###

https://tools.hmcts.net/jira/browse/FACT-539

### Change description ###

- Prevent duplicated opening hours types being adding to admin portal
- Add the ability to return multiple errors in error summary

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/test/functional/features/opening-hours.feature [hunk 1]
```diff
@@ -21,15 +21,20 @@ Feature: Opening Hours
...
-    Then an error message is displayed
-
   Scenario: Remove opening hours
     When I click the remove button under an opening hours entry
     Then I click save
     Then a green update message is displayed in the opening hours tab
     When I click the remove button under an opening hours entry
     Then I click save
     Then a green update message is displayed in the opening hours tab
+
```
Comment (agraham01) [inline src/test/functional/features/opening-hours.feature:31]: The new duplicate checking might cause an issue for the existing functional tests. If the functional tests fail after the test where the new opening hours are added and before the test where the new additions are deleted, there is a possibility a duplicate will occur next time the tests are run. This will likely cause the 'add new opening hours' test to fail when the new entry isn't saved due to the duplicate.

File: src/test/unit/app/controller/courts/OpeningTimesController.ts [hunk 2]
```diff
@@ -135,7 +152,7 @@ describe('OpeningTimesController', () => {
       'opening_times': postedOpeningTimes,
       openingTimeTypes: expectedSelectItems,
       updated: false,
-      errorMsg: 'A problem occurred when saving the opening times.'
+      errors: [{text: 'A problem occurred when saving the opening times.'}]
```
Comment (agraham01) [inline src/test/unit/app/controller/courts/OpeningTimesController.ts:155]: This error message exists as a private field in the controller class. I wonder if we should make that public and just refer to the variable here to avoid the test failing if we ever want to change the text.

File: src/test/functional/features/opening-hours.feature (hunk 1 above)
Comment (KianKwa) [reply src/test/functional/features/opening-hours.feature:31]: Currently if the “add opening hours” test fails after adding the opening hours, I thought it still carries on to the “remove opening hours” test? The problem can occur when we only manage to remove one opening hour in the “remove opening hours” test and the test somehow fails after that, resulting in one of the entries not getting removed. The next time we run the “add opening hours” test we might have the problem with duplicates, but the “remove opening hours” test should remove the remaining entry from the first run. So the third run should all be successful. This is not ideal.

We could remove all the entries before starting each test case. We know that the Aberdare County Court has no opening hours, phone number, emails and court types before the functional tests. For the opening hours test, we could add a step at the end of the background step to remove all added entries before starting each test. We can do that in such a way that we count the number of fieldsets before the removal and remove all the present fields (if any), so that the form is empty before starting any new test scenario.

If we do this, we don’t really need the explicit “remove opening hours” scenario as this will be tested as part of the background steps. If we want to keep that scenario, we will need to combine with the ‘add opening hours’ scenario and rename them “Add and remove opening hours” so they are tested together.

Let me know what you think.

File: src/test/unit/app/controller/courts/OpeningTimesController.ts (hunk 2 above)
Comment (KianKwa) [reply src/test/unit/app/controller/courts/OpeningTimesController.ts]: I have made the test using the error string from the controller class

File: src/test/functional/utlis/helpers.ts [hunk 3]
```diff
@@ -15,4 +15,26 @@ export class FunctionalTestHelpers {
...
+
+  public static async clearFieldsets(containerId: string, buttonName: string) {
+    const fieldsetSelector = `${containerId} fieldset`;
+    const fieldsetCount = await I.countElement(fieldsetSelector);
+    // Remove all fieldsets except the empty new one and the template
+    for (let i = fieldsetCount; i > 2; i--) {
+      const deleteButtonSelector = `${containerId} button[name="${buttonName}"]`;
+      const elementExist = await I.checkElement(deleteButtonSelector);
+      expect(elementExist).equal(true);
+      await I.click(deleteButtonSelector);
```
Comment (agraham01) [inline src/test/functional/utlis/helpers.ts:27]: Could you call your new clickSaveButton method here?

File: src/test/functional/utlis/helpers.ts [hunk 4]
```diff
@@ -15,4 +15,26 @@ export class FunctionalTestHelpers {
...
+
+      const updatedFieldsetCount = await I.countElement(fieldsetSelector);
+      expect(i - updatedFieldsetCount).equal(1);
+    }
+  }
+
+  public static async clickSaveButton(containerId: string, buttonName: string) {
```
Comment (agraham01) [inline src/test/functional/utlis/helpers.ts:34]: Should we rename this to remove 'save' - it's really clicking any button with the given button name

File: src/test/functional/utlis/helpers.ts (hunk 3 above)
Comment (KianKwa) [reply src/test/functional/utlis/helpers.ts:27]: I have changed it to call 'save' directly from the helper method

**Review by agraham01 — APPROVED** [review]
